src/VisualStudio/Core/Def/Progression/GraphBuilder.cs

```diff
@@ -76,22 +76,30 @@ private void PopulateMapsForFileInputNode(GraphNode inputNode, CancellationToken
                     return;
                 }
 
-                var project = _solution.Projects.FirstOrDefault(
-                    p => string.Equals(p.FilePath, projectPath.OriginalString, StringComparison.OrdinalIgnoreCase));
-                if (project == null)
+                // Don't walk _solution.Projects as that will possibly create projects that don't match
+                // our criteria. Instead, walk the project states to find the appropriate project id.
+                var projectState = _solution.SolutionState.ProjectStates.FirstOrDefault(
+                    p => string.Equals(p.Value.FilePath, projectPath.OriginalString, StringComparison.OrdinalIgnoreCase));
+                if (projectState.Key == null)
                 {
                     return;
                 }
 
+                var project = _solution.GetRequiredProject(projectState.Key);
+
                 _nodeToContextProjectMap.Add(inputNode, project);
 
-                var document = project.Documents.FirstOrDefault(
-                    d => string.Equals(d.FilePath, filePath.OriginalString, StringComparison.OrdinalIgnoreCase));
-                if (document == null)
+                // Don't walk project.Documents as that will possibly create documents that don't match
+                // our criteria. Instead, walk the document states to find the appropriate document id.
+                var documentState = project.State.DocumentStates.States.FirstOrDefault(
+                    d => string.Equals(d.Value.FilePath, filePath.OriginalString, StringComparison.OrdinalIgnoreCase));
+                if (documentState.Key == null)
```

Comment:
What might be better still: just call [GetDocumentIdsWithFilePath(String)](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.solution.getdocumentidswithfilepath?view=roslyn-dotnet-4.7.0#microsoft-codeanalysis-solution-getdocumentidswithfilepath(system-string)), and if we get more than one then compare the project's paths. That way we'll significantly cut down on string comparisons.

Reply:
Ooh, I like that. Let me give that a whirl...

Reply:
Yeah, much nicer. Cleaner and more efficient!

Reply:
test is failing, not obvious why. I'll dig in more in the morning.